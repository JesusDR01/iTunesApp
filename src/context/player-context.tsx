import { Podcast, Podcasts } from '@modules/podcasts/domain/Podcast';
import {
	Dispatch,
	MutableRefObject,
	createContext,
	useContext,
	useReducer,
	useRef,
} from 'react';

export type PodcastList = Podcasts;


export type State = {
	randomMode: boolean;
	currentPodcast: Podcast | null;
	podcastList?: PodcastList;
	isPlaying: boolean;
	volume: number;
	time: number;
	loopMode: 'all' | 'one' | 'none';
};

export type Action =
	| { type: 'play' }
	| { type: 'pause' }
	| { type: 'setPodcast'; podcast: Podcast }
	| { type: 'setPodcastList'; podcastList: PodcastList }
	| { type: 'loopMode'; loopMode: 'all' | 'one' | 'none' }
	| { type: 'randomMode'; randomMode: boolean }
	| { type: 'next' }
	| { type: 'end' }
	| { type: 'prev' }
	| { type: 'volume'; volume: number };

export type DispatchAction = React.Dispatch<Action>;

interface SearchPodcastContext {
	audioRef: MutableRefObject<HTMLAudioElement | null>;
	useControls: [State, Dispatch<Action>];
}

export const PlayerContext = createContext<SearchPodcastContext | null>(null);

export const PlayerContextProvider = ({
	children,
}: {
	children: React.ReactNode;
}) => {
	const audioRef = useRef<HTMLAudioElement | null>(null);
	const getInitialState = (): State => ({
		randomMode: false,
		currentPodcast: null,
		podcastList: undefined,
		isPlaying: false,
		volume: 0.3,
		time: 0,
		loopMode: 'none',
	});

	return (
		<PlayerContext.Provider
			value={{
				audioRef,
				useControls: useReducer((state: State, action: Action) => {
					switch (action.type) {
						case 'setPodcast': {
							const podcast = action.podcast;
							return {
								...state,
								currentPodcast: podcast,
							};
						}
						case 'setPodcastList': {
							const podcastList = action.podcastList;
							return {
								...state,
								podcastList: podcastList,
							};
						}
						case 'loopMode': {
							return { ...state, loopMode: action.loopMode };
						}
						case 'randomMode': {
							const randomMode = action.randomMode;
							return {
								...state,
								randomMode: randomMode,
							};
						}
						case 'prev': {
							const { podcastList, currentPodcast, loopMode } = state;
							if (loopMode === 'one' && audioRef.current) {
								audioRef.current.currentTime = 0;
								return {
									...state,
								};
							}
							if (podcastList && currentPodcast) {
								const currentPodcastIndex = podcastList.findIndex(
									podcast => podcast.episodeUrl === currentPodcast?.episodeUrl,
								);
								const prevPodcast = podcastList[currentPodcastIndex - 1];

								return {
									...state,
									currentPodcast: prevPodcast || currentPodcast,
								};
							}

							return {
								...state,
							};
						}
						case 'next': {
							const { podcastList, currentPodcast, randomMode, loopMode } =
								state;

							if (podcastList && currentPodcast) {
								const currentPodcastIndex = podcastList.findIndex(
									podcast => podcast.episodeUrl === currentPodcast?.episodeUrl,
								);

								const nextPodcast =
									podcastList[
										randomMode
											? Math.floor(Math.random() * podcastList.length)
											: currentPodcastIndex + 1
									];
								return {
									...state,
									currentPodcast: nextPodcast || currentPodcast,
									loopMode:
										loopMode === 'one' || loopMode === 'all'
											? ('all' as const)
											: ('none' as const),
								};
							}

							return {
								...state,
							};
						}
						case 'end': {
							const { podcastList, currentPodcast, randomMode, loopMode } =
								state;

								if (loopMode === 'one' && audioRef.current) {
									audioRef.current.currentTime = 0;
									audioRef.current.play();
									return {
										...state,
									};
								}

							if (podcastList && currentPodcast) {
								const currentPodcastIndex = podcastList.findIndex(
									podcast => podcast.episodeUrl === currentPodcast?.episodeUrl,
								);

								const nextPodcast =
									podcastList[
										randomMode
											? Math.floor(Math.random() * podcastList.length)
											: currentPodcastIndex + 1
									];
								return {
									...state,
									currentPodcast: nextPodcast || currentPodcast,
								};
							}

							return {
								...state,
							};
						}
						case 'play': {
							audioRef.current?.play();
							return { ...state, isPlaying: true };
						}
						case 'pause': {
							audioRef.current?.pause();
							return { ...state, isPlaying: false };
						}
						case 'volume': {
							const volume = action.volume;
							if (audioRef.current) {
								audioRef.current.volume = volume;
							}
							return {
								...state,
								volume: volume,
							};
						}
						default: {
							return state;
						}
					}
				}, getInitialState()),
			}}
		>
			{children}
		</PlayerContext.Provider>
	);
};

export const usePlayer = () => {
	const context = useContext(PlayerContext);

	if (context === null) {
		throw new Error('SearchPodcastContext error');
	}

	return context;
};
