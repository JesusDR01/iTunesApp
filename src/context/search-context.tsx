import {
	Dispatch,
	SetStateAction,
	createContext,
	useContext,
	useState,
} from 'react';

interface SearchPodcastContext {
	useSearch: [string, Dispatch<SetStateAction<string>>];
}

export const SearchPodcastContext = createContext<SearchPodcastContext | null>(
	null,
);

export const SearchPodcastsProvider = ({
	children,
}: {
	children: React.ReactNode;
}) => {
	const [search, setSearch] = useState('');

	return (
		<SearchPodcastContext.Provider
			value={{
				useSearch: [search, setSearch],
			}}
		>
			{children}
		</SearchPodcastContext.Provider>
	);
};

export const useSearch = () => {
	const context = useContext(SearchPodcastContext);

	if (context === null) {
		throw new Error('SearchPodcastContext error');
	}

	return context;
};
