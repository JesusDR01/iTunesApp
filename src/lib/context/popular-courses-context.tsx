import {
  Dispatch,
  SetStateAction,
  createContext,
  useContext,
  useState,
} from "react"

interface PopularCoursesContext {
  useDefaultValue: [string, Dispatch<SetStateAction<string>>]
}

export const PopularCoursesContext =
  createContext<PopularCoursesContext | null>(null)

export const PopularCoursesProvider = ({
  children,
}: {
  children: React.ReactNode
}) => {
  const [screen, setScreen] = useState("")
  return (
    <PopularCoursesContext.Provider
      value={{
        useDefaultValue: [screen, setScreen],
      }}
    >
      {children}
    </PopularCoursesContext.Provider>
  )
}

export const usePopularCourses = () => {
  const context = useContext(PopularCoursesContext)

  if (context === null) {
    throw new Error("PopularCoursesContext error")
  }

  return context
}
