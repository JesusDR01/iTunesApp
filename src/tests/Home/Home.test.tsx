import { render } from '@testing-library/react';

import HomePage from 'pages/index';

import { QueryClient, QueryClientProvider } from '@tanstack/react-query';
import { SearchPodcastsProvider } from 'context/search-context';
import { PlayerContextProvider } from 'context/player-context';

const queryClient = new QueryClient({
	defaultOptions: {
		queries: {
			retry: false,
		},
	},
});

const AllTheProviders = ({ children }: { children: React.ReactElement }) => {
	return (
		<QueryClientProvider client={queryClient}>
			<PlayerContextProvider>
				<SearchPodcastsProvider>{children}</SearchPodcastsProvider>
			</PlayerContextProvider>
		</QueryClientProvider>
	);
};

// const filters = [
// 	{ type: 'sport', value: 'football', name: 'Football' },
// 	{ type: 'competition', value: 'premier-league', name: 'Premier League' },
// ];

// const mockUseFilters = {
// 	filters,
// 	removeFilter: jest.fn(),
// 	addFilter: jest.fn(),
// 	loading: false,
// };
// jest.mock('hooks/use-stored-filters', () => ({
// 	useStoredFilters: () => mockUseFilters,
// }));

// const useScheduleData = {
// 	isLoading: false,
// 	dates: [],
// 	events: [],
// 	update: jest.fn(),
// };

// jest.mock('./hooks/use-schedule', () => () => useScheduleData);

describe('Home screen', () => {
	it('should render the home screen without crashing', () => {
		const { container } = render(<HomePage />, { wrapper: AllTheProviders });
		expect(container).toBeInTheDocument();
	});

	it('should match snapshot', () => {
		const { baseElement } = render(<HomePage />, { wrapper: AllTheProviders });
		expect(baseElement).toMatchSnapshot();
	});

	// it('should check filters', () => {
	// 	const { getByTestId } = render(<HomePage />, { wrapper: AllTheProviders });

	// 	expect(getByTestId(ID_FILTERS)).toHaveTextContent(filters[0].name);
	// 	expect(getByTestId(ID_FILTERS)).toHaveTextContent(filters[1].name);
	// });

	// it('should remove filter and focus the next one', async () => {
	// 	const { getByTestId, getByText, rerender } = render(<HomePage />, {
	// 		wrapper: AllTheProviders,
	// 	});

	// 	mockUseFilters.removeFilter.mockImplementation(filter => {
	// 		const idx = filters.findIndex(
	// 			f => f.type === filter.type && f.value === filter.value,
	// 		);
	// 		idx !== -1 && filters.splice(idx, 1);
	// 	});

	// 	const firstFilter = mockUseFilters.filters[0];
	// 	await fireEvent.click(getByText(firstFilter.name));
	// 	rerender(<HomePage />);
	// 	expect(mockUseFilters.removeFilter).toHaveBeenCalled();
	// 	expect(getByTestId(ID_FILTERS)).not.toHaveTextContent(firstFilter.name);
	// });
});
