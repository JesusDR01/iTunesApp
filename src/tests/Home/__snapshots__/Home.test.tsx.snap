// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home screen should match snapshot 1`] = `
<body>
  <div>
    <main
      class="className"
    >
      <div
        class="MuiGrid-root MuiGrid-container MuiGrid-direction-xs-column min-h-[70vh] css-l6hioc-MuiGrid-root"
        style="min-height: 70vh;"
      >
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-3 css-1equabv-MuiGrid-root"
        >
          <span
            class="MuiCircularProgress-root MuiCircularProgress-indeterminate MuiCircularProgress-colorPrimary !text-white css-18lrjg1-MuiCircularProgress-root"
            role="progressbar"
            style="width: 40px; height: 40px;"
          >
            <svg
              class="MuiCircularProgress-svg css-1idz92c-MuiCircularProgress-svg"
              viewBox="22 22 44 44"
            >
              <circle
                class="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate css-176wh8e-MuiCircularProgress-circle"
                cx="44"
                cy="44"
                fill="none"
                r="20.2"
                stroke-width="3.6"
              />
            </svg>
          </span>
        </div>
      </div>
    </main>
  </div>
</body>
`;
